internal/driver/driver.go

```diff
@@ -0,0 +1,793 @@
+	"github.com/edgexfoundry/go-mod-core-contracts/v3/clients/logger"
+	"github.com/edgexfoundry/go-mod-core-contracts/v3/common"
+	"github.com/edgexfoundry/go-mod-core-contracts/v3/models"
+	"github.com/robinson/gos7"
+	"github.com/spf13/cast"
+)
+
+const readCommandsExecutedName = "ReadCommandsExecuted"
+
+const (
+	// Word Length
+	s7wlbit     = 0x01 //Bit (inside a word)
+	s7wlbyte    = 0x02 //Byte (8 bit)
+	s7wlChar    = 0x03
+	s7wlword    = 0x04 //Word (16 bit)
+	s7wlint     = 0x05
+	s7wldword   = 0x06 //Double Word (32 bit)
+	s7wldint    = 0x07
+	s7wlreal    = 0x08 //Real (32 bit float)
+	s7wlcounter = 0x1C //Counter (16 bit)
+	s7wltimer   = 0x1D //Timer (16 bit)
+)
+
+var once sync.Once
+var driver *Driver
+
+type Driver struct {
+	lc                   logger.LoggingClient
+	asyncCh              chan<- *sdkModel.AsyncValues
+	stringArray          []string
+	serviceConfig        *ServiceConfig
+	sdkService           interfaces.DeviceServiceSDK
+	s7Clients            map[string]*S7Client
+}
+
+func NewProtocolDriver() interfaces.ProtocolDriver {
+	once.Do(func() {
+		driver = new(Driver)
+	})
+	return driver
+}
+
+type CommandInfo struct {
+	Host            string
+	Port            int
+	Rack            int
+	Slot            int
+	AddressType     string
+	DbAddress       int
+	StartingAddress int
+	Length          int
+	Pos             int
+	ValueType       string
+}
+
+type DBInfo struct {
+	Area       int
+	DBNumber   int
+	Start      int
+	Amount     int
+	WordLength int
+	DBArray    []string
+}
+
+// Initialize performs protocol-specific initialization for the device
+// service.
+func (s *Driver) Initialize(sdk interfaces.DeviceServiceSDK) error {
+	s.lc = sdk.LoggingClient()
+	s.asyncCh = sdk.AsyncValuesChannel()
+	s.serviceConfig = &ServiceConfig{}
+	s.s7Clients = make(map[string]*S7Client)
+
+	s.stringArray = []string{"foo", "bar"}
+
+	// ds := interfaces.Service()
+	// s.sdkService = service.RunningService()
+	// ds := service.RunningService()
+
+	// if err := ds.LoadCustomConfig(s.serviceConfig, CustomConfigSectionName); err != nil {
+	// 	return fmt.Errorf("unable to load 'SimpleCustom' custom configuration: %s", err.Error())
+	// }
+
+	// s.lc.Infof("Custom config is: %v", s.serviceConfig)
+
+	// if err := s.serviceConfig.S7Info.Validate(); err != nil {
+	// 	return fmt.Errorf("'S7InfoCustom' custom configuration validation failed: %s", err.Error())
+	// }
+
+	// if err := ds.ListenForCustomConfigChanges(
+	// 	&s.serviceConfig.S7Info.Writable,
+	// 	WritableInfoSectionName, s.ProcessCustomConfigChanges); err != nil {
+	// 	return fmt.Errorf("unable to listen for changes for 'S7Info.Writable' custom configuration: %s", err.Error())
+	// }
+
+	for _, device := range sdk.Devices() {
+		s7Client := s.NewS7Client(device.Name, device.Protocols)
+		if s7Client == nil {
+			s.lc.Errorf("failed to initialize S7 client for '%s' device, skipping this device.", device.Name)
+			continue
+		}
+		s.s7Clients[device.Name] = s7Client
+	}
```

Comment:
This should be done in AddDevice() when each device is added.

Reply:
For initialize function, should I do this?

Reply:
No you can now remove it,